File: homework/ba-nguyen/lesson-8/two-sum-bsts.ts
function twoSumBSTs(
  root1: TreeNode | null,
  root2: TreeNode | null,
  target: number
): boolean {
  const queue: TreeNode[] = [];
  queue.push(root1);

  while (!!queue.length) {
    const length = queue.length;
    for (let i = 0; i < length; i++) {
      const node = queue.shift();
      if (binarySearch(root2, target - node.val)) return true;
      if (node.left) queue.push(node.left);
      if (node.right) queue.push(node.right);
    }
  }

  return false;
}

function binarySearch(node: TreeNode, numb: number) {
  if (!node) return false;

  if (node.val === numb) return true;
  if (node.val < numb) return binarySearch(node.right, numb);
  return binarySearch(node.left, numb);
}

File: homework/Long_Nguyen/lesson_8/TwoSumBSTs.java
import java.util.HashSet;
import java.util.Set;

/**
 * Created by Long Nguyen
 */
public class TwoSumBSTs {
    public boolean twoSumBSTs(TreeNode root1, TreeNode root2, int target) {
        Set<Integer> cache = new HashSet<>();
        inOrder(root1, target, cache);
        return check(root2, cache);
    }

    public void inOrder(TreeNode root, int target, Set<Integer> cache) {
        if (root == null)
            return;
        inOrder(root.left, target, cache);
        cache.add(target - root.val);
        inOrder(root.right, target, cache);
    }

    public boolean check(TreeNode root, Set<Integer> cache) {
        if (root == null)
            return false;
        if (cache.contains(root.val))
            return false;
        if (check(root.left, cache))
            return true;
        if (check(root.right, cache))
            return true;

        return false;
    }
}
